PostCorrecter

Bobritta opened the pull request.

Commit:
PostCorrecter

Commit:
an attempt to split a large method into functional parts

Razzzek commented on src/main/java/faang/school/postservice/config/corrector/PostCorrecter.java:
```diff
@@ -0,0 +1,81 @@
+package faang.school.postservice.config.corrector;
```
Можно сказать, что твой класс является сервисом, поэтому его лучше переместить в соответствующий пакет. Возможно стоит подумать о создании отдельного интерфейса с последующей имплементацией

Razzzek commented on src/main/java/faang/school/postservice/config/corrector/PostCorrecter.java:
```diff
@@ -0,0 +1,81 @@
+        StringBuilder correctedText = new StringBuilder();
+        try {
+            String url = properties.apiUrl();
+            HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
```
Предлагаю тебе попробовать RestTemplate вместо HttpURLConnection, он современный и его гораздо проще настроить

Razzzek commented on src/main/java/faang/school/postservice/config/corrector/PostCorrecter.java:
```diff
@@ -0,0 +1,81 @@
+        }
+        input.close();
+
+        ObjectMapper objectMapper = new ObjectMapper();
```
ObjectMapper лучше сделать бином спринга и внедрить его в данный сервис, чтобы не создавать каждый раз

Razzzek commented on src/main/java/faang/school/postservice/config/corrector/PostCorrecter.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private static JsonNode getJsonNode(HttpURLConnection connection) throws IOException {
+        BufferedReader input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
```
С RestTemplate тебе не придется создавать свой reader и закрывать его, эти функции работают "из коробки"

Razzzek reviewed (changes requested):
Привет! Здесь кратко расписал, подробнее в дискорде напишу

Commit:
before changing HttpURLConnection to RestTemplate

Commit:
HttpURLConnection has been changed to RestTemplate

Commit:
fixed

Commit:
improved

Commit:
брала лишний импорт

Commit:
брала лишний импорт и пустые строки

Commit:
брала лишний импорт и пустые строки

Commit:
убрала лишний @ConfigurationPropertiesScan

Commit:
after rebasing

miked4king commented on src/main/java/faang/school/postservice/dto/post/TextCheckRequest.java:
```diff
@@ -0,0 +1,14 @@
+@NoArgsConstructor
+public class TextCheckRequest {
+    @NotBlank(message = "Post content could not be blank")
+    String text;
```
модификатор доступа потеряли)

miked4king commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -125,4 +128,11 @@ public List<PostOutputDto> getNotDeletedProjectPublished(@NotNull @PathVariable(
+
+    @PostMapping("/check-text")
+    public TextCheckRequest checkPostText(@RequestBody TextCheckRequest request) {
+        String textToCheck = request.getText();
```
можем не выносить в отдельную переменную, он же у нас в одном месте используется)

miked4king commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -125,4 +128,11 @@ public List<PostOutputDto> getNotDeletedProjectPublished(@NotNull @PathVariable(
+    public TextCheckRequest checkPostText(@RequestBody TextCheckRequest request) {
+        String textToCheck = request.getText();
+        String correctedText = postCorrecter.checkText(textToCheck);
+        return new TextCheckRequest(correctedText);
```
может для ответа отдельную dto сделаем и сразу из сервиса будем возвращать?

Bobritta replied:
две ДТО с единственным полем String text для одного эндпоинта, так и правда нужно?

miked4king commented on src/main/java/faang/school/postservice/service/post/PostCorrecterImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    @Retryable(retryFor = {IOException.class}, maxAttempts = 3, backoff = @Backoff(delay = 2000, multiplier = 1.5))
+    public String checkText(String textToCheck) {
```
немного каши в методе, давай вынесем вызов внешнего сервиса в отдельный компонент.
в данном сервисе мы будем подготавливать запрос и обрабатывать ответ

miked4king commented on src/main/java/faang/school/postservice/service/post/PostCorrecterImpl.java:
```diff
@@ -0,0 +1,62 @@
+                        new ParameterizedTypeReference<JsonNode>() {
+                        });
+
+        if (responseEntity.getStatusCode().is2xxSuccessful()) {
```
при 4xx/5xx ответах будет выбрасываться исключение, можем через try catch обработать
можешь тут посмотреть как обрабатывать исключения https://hellokoding.com/spring-resttemplate-error-handling/

miked4king commented on src/main/java/faang/school/postservice/service/post/PostCorrecterImpl.java:
```diff
@@ -0,0 +1,62 @@
+        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+        String params = "text=" + URLEncoder.encode(textToCheck, StandardCharsets.UTF_8) + "&language=" + properties.language();
+        HttpEntity<String> requestEntity = new HttpEntity<>(params, headers);
+        ResponseEntity<JsonNode> responseEntity =
```
а нам обязательно JsonNode принимать? java объект может быть получится?

Commit:
fix

Commit:
divided PostCorrecterImpl and TextCorrectionService

Commit:
CorrectionResponse instead of JsonNode

Commit:
RestClientResponseException handling

Commit:
responseDto

miked4king commented on src/main/java/faang/school/postservice/service/text/TextCorrectionService.java:
```diff
@@ -0,0 +1,43 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class TextCorrectionService {
```
1) забыли добавить интерфейс)
2) давай имя класса поправим, все таки это компонент осуществляющий вызовы внешней системы, хочется видеть в названии слово client (как в других репозиториях) или gateway
3) давай вынесем из пакета service, у вас в других сервисах используется пакет client

miked4king commented on src/main/java/faang/school/postservice/service/text/TextCorrectionService.java:
```diff
@@ -0,0 +1,43 @@
+    private final RestTemplate restTemplate;
+
+    public CorrectionResponse callCorrectionApi(String params) {
+        String url = properties.apiUrl();
```
можем не записывать в переменную с учетом что один раз используется

miked4king commented on src/main/java/faang/school/postservice/service/text/TextCorrectionService.java:
```diff
@@ -0,0 +1,43 @@
+            return responseEntity.getBody();
+        } catch (RestClientResponseException e) {
+            log.error("Failed : HTTP error while post text autocorrecting. Text & language = {}", params, e);
+            throw new TextAutoCorrectionException("Failed : HTTP error code : " + ResponseEntity
```
она обрабатывается в exception handler?
сообщение через String.format

Bobritta replied:
я не видела смысла её отдельно обрабатывать в error  handler, потому что не будет для неё специфического статуса или сообщения. Т.е. она пролетит все сравнения и обработается в последнем:
 @ExceptionHandler(Exception.class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    public ErrorResponseDto handleException(Exception e) {
        log.error("Exception was thrown", e);
        return new ErrorResponseDto(
                HttpStatus.INTERNAL_SERVER_ERROR.name(),
                "Something get wrong.",
                e.getMessage(),
                LocalDateTime.now().format(formatter)
        );
        Тут же нет вещей, о которых нам нужно сказать клиентам: это либо мы неправильно передаём, либо на стороне текст-редактора косяк

Bobritta replied:
или есть причины?

miked4king commented on src/main/java/faang/school/postservice/service/post/PostCorrecterImpl.java:
```diff
@@ -0,0 +1,45 @@
+    private final TextCorrectionService textCorrectionService;
+
+    @Override
+    @Retryable(retryFor = {IOException.class}, maxAttempts = 3, backoff = @Backoff(delay = 2000, multiplier = 1.5))
```
эту аннотацию думаю лучше повесить на метод, который отправляет запрос во внешнюю систему.
кстати, почему retry на IOException? в каком случае она у нас тут будет?
плюс параметры  maxAttempts = 3, backoff = @Backoff(delay = 2000, multiplier = 1.5 лучше вынести в конфигурацию, чтобы можно было без изменения кода конфигурировать

miked4king commented on src/main/java/faang/school/postservice/service/post/PostCorrecterImpl.java:
```diff
@@ -0,0 +1,45 @@
+        String params = "text=" + URLEncoder.encode(textToCheck, StandardCharsets.UTF_8)
+                + "&language=" + properties.language();
+        CorrectionResponse correctionResponse = textCorrectionService.callCorrectionApi(params);
+            int previousEnd = 0;
+            assert correctionResponse != null;
+            for (CorrectionResponse.Match match : correctionResponse.getMatches()) {
+                if (match.getReplacements() != null && !match.getReplacements().isEmpty()) {
+                    String replacement = match.getReplacements().get(0).getValue();
+                    int offset = match.getOffset();
+                    int length = match.getLength();
+                    correctedText.append(textToCheck, previousEnd, offset);
+                    correctedText.append(replacement);
+                    previousEnd = offset + length;
+                }
+            }
+            correctedText.append(textToCheck.substring(previousEnd));
```
этот блок кажется ушел вправо по форматированию

miked4king commented on src/main/java/faang/school/postservice/service/post/PostCorrecterImpl.java:
```diff
@@ -0,0 +1,45 @@
+import java.nio.charset.StandardCharsets;
+
+@RequiredArgsConstructor
+@Component
```
@Service

Commit:
RetryProps

miked4king commented on src/main/java/faang/school/postservice/client/TextCorrectionClient.java:
```diff
@@ -0,0 +1,7 @@
+import faang.school.postservice.model.text.CorrectionResponse;
+
+public interface TextCorrectionClient {
+    public CorrectionResponse callCorrectionApi(String params);
```
public - по умолчанию)

miked4king commented on src/main/java/faang/school/postservice/client/text/TextCorrectionClientImpl.java:
```diff
@@ -0,0 +1,49 @@
+import java.io.IOException;
+
+@Slf4j
+@Service
```
@Component

miked4king commented on src/main/java/faang/school/postservice/client/text/TextCorrectionClientImpl.java:
```diff
@@ -0,0 +1,49 @@
+    private final PostCorrectorProperty properties;
+    private final RestTemplate restTemplate;
+
+    @Retryable(retryFor = {RestClientResponseException .class},
```
.

Commit:
убрала текстАвтокоррекшнЭксепшн, теперь будут улавливаться все виды RestClientResponseException, добавила обработку той части, что выкидывает 4хх ответы в хендлер

Commit:
не вышло отказаться от класса с конфигурацией для аннотации retryable. Пробовала создавать внутренний класс в
TextCorrectionClientImpl или использовать в нём @Value для прямой инициализации полей, но нет:
@Retryable не поддерживает прямое использование полей класса для параметров maxAttemptsExpression и backoff. Вместо этого, вы должны использовать Spring Expression Language (SpEL) для указания значений, которые должны быть разрешены в контексте Spring

Bobritta commented:
убрала текстАвтокоррекшнЭксепшн, теперь будут улавливаться все виды RestClientResponseException, добавила обработку той части, что выкидывает 4хх ответы в хендлер 
не вышло отказаться от класса с конфигурацией для аннотации retryable. Пробовала создавать внутренний класс в TextCorrectionClientImpl или использовать в нём @Value для прямой инициализации полей, но нет: @Retryable не поддерживает прямое использование полей класса для параметров maxAttemptsExpression и backoff. Вместо этого, вы должны использовать Spring Expression Language (SpEL) для указания значений, которые должны быть разрешены в контексте Spring

miked4king commented:
Принимаю задачу) 
разреши конфликты в application.yaml и можем мержить

Commit:
Merge branch 'kelpie-master-stream10' into develop/BJS2-77662

# Conflicts:
#	src/main/resources/application.yaml
